include/nlohmann/json.hpp

```diff
@@ -4707,20 +4707,16 @@ std::string to_string(const NLOHMANN_BASIC_JSON_TPL& j)
 // nonmember support //
 ///////////////////////
 
-// specialization of std::swap, and std::hash
-namespace std
-{
-
 /// hash value for JSON objects
-template<>
-struct hash<nlohmann::json>
+template<typename BasicJsonType>
+struct std::hash
```

Comment:
@nlohmann This is now a full redefinition of the `std::hash` struct, which is not allowed.  You still need to keep it limited to the library's type.  It needs to stay as before but use `nlohmann::basic_json<>>` instead of `nlohmann::json`.  (I'm not sure the exact proper way to spell `nlohmann::basic_json<>` here, I would start with that and see if it works.  If not, you'll have to add more template parameters.)

Reply:
I think you need to do it like this:

    namespace std
    {
       NLOHMANN_BASIC_JSON_TPL_DECLARATION
       struct hash<NLOHMANN_BASIC_JSON_TPL>
       {
           std::size_t operator()(const NLOHMANN_BASIC_JSON_TPL &j)

Reply:
Thanks a lot! I actually saw your fix before I noticed the CI was red ;-)

I will check out the other functions as well - it seems just having `nlohmann::json` in the code base usually is a problem in the future, as it will not work for other specializations like `nlohmann::ordered_json`.

Reply:
I didn't notice the CI was red.  :)  I wasn't sure if this was something that the CI would catch, or if it would just lead to a silent ODR violation later.

Reply:
I recently read https://quuxplusone.github.io/blog/2021/10/27/dont-reopen-namespace-std/ and wanted to apply the ideas. With that, I notices we lacked the specialization for `nlohmann::ordered_json`. Interestingly though, GCC 4.8 rejects the approach of the blog post, so we're back to the original code, but with your fix to the specialization.

Reply:
Yeah, looks like this was a GCC bug until 7.x.